workbench-core/environments/src/handlers/environmentTypeHandler.ts

```diff
@@ -0,0 +1,106 @@
+/*
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *  SPDX-License-Identifier: Apache-2.0
+ */
+
+import { ProductViewSummary, ProvisioningArtifactDetail } from '@aws-sdk/client-service-catalog';
+import { AwsService, resourceTypeToKey, CFNTemplateParameters } from '@aws/workbench-core-base';
+import { EnvironmentType } from '../models/environmentTypes/environmentType';
+import EnvironmentTypeService from '../services/environmentTypeService';
+
+export default class EnvironmentTypeHandler {
+  private _mainAccountAwsService: AwsService;
+
+  public constructor(mainAccountAwsService: AwsService) {
+    this._mainAccountAwsService = mainAccountAwsService;
+  }
+  /* eslint-disable-next-line @typescript-eslint/no-explicit-any*/
+  public async execute(event: any): Promise<void> {
+    console.log(`Processing Environment Types`);
+    const portfolioName = process.env.SC_PORTFOLIO_NAME!;
+    const portfolioId = await this._mainAccountAwsService.helpers.serviceCatalog.getPortfolioId(
+      portfolioName
+    );
+    if (portfolioId === undefined) {
+      throw new Error(`Could not find portfolioId for portfolio: ${portfolioName}`);
+    }
+    console.log(`Processing portfolio ${portfolioId}`);
+    const products = await this._mainAccountAwsService.helpers.serviceCatalog.getProductsByPortfolioId(
+      portfolioId
+    );
+    for (const product of products) {
+      try {
+        console.log(`Processing product ${product.ProductId}`);
+        const currentProductId = product.ProductId || '';
+        const provisionArtifacts =
+          await this._mainAccountAwsService.helpers.serviceCatalog.getProvisionArtifactsByProductId(
+            currentProductId
+          );
+        for (const provisionArtifact of provisionArtifacts.filter((p) => p.Active)) {
+          try {
+            console.log(`Processing provision artifact ${provisionArtifact.Id}-${product.ProductId}`);
+            const details =
+              await this._mainAccountAwsService.helpers.serviceCatalog.getProvisionArtifactDetails(
+                currentProductId,
+                provisionArtifact.Id || ''
+              );
+            const envTypeId = `${resourceTypeToKey.envType.toLowerCase()}-${product.ProductId},${
+              provisionArtifact.Id
+            }`;
+            const existingEnvType = await this._getExistingEnvironmentType(envTypeId);
+            if (!existingEnvType && !!details?.Info?.TemplateUrl) {
+              console.log(`Creating environment type: ${envTypeId}.`);
+              const params = await this._mainAccountAwsService.helpers.s3.getTemplateByURL(
+                details.Info.TemplateUrl
+              );
+              await this._saveEnvironmentType(product, provisionArtifact, params?.Parameters);
+            }
+          } catch (e) {
+            console.log(
+              `An error ocurred while trying to process Provision Artifact: ${provisionArtifact.Id} - ${e}`
```

Comment:
do you need to `JSON.stringify(e)` the error for it to show up as pretty-printed JSON?

Reply:
fixing next commit

Reply:
replaced with `JSON.stringify`